runtime/compiler/optimizer/EscapeAnalysisTools.cpp

```diff
@@ -153,6 +202,18 @@ void TR_EscapeAnalysisTools::processSymbolReferences(TR_Array<List<TR::SymbolRef
                       && (deadSymRefs == NULL || !deadSymRefs->isSet(definingSymRefNum)))
                      {
                      _loads->push_back(TR::Node::createWithSymRef(TR::aload, 0, definingSymRef));
+
```

Comment:
Are you able to explain the difference between these two affected clauses in `processSymbolReferences` ? i.e. what exactly does a sym ref being in the `induceDefiningMap` mean at this point ? Is it akin to that sym ref having been stored to earlier in the method, i.e. a definition of the sym ref is reaching this specific program point or is it independent of the program point, i.e. bytecode index we are inducing at ? This may depend on the answer to the prior question, but I was wondering if the same test would be valid for a parm sym ref (that may not have been stored explicitly).

Reply:
It took me a while to reconstruct how this works.  My understanding is that if a symref is in the `induceDefiningMap`, that means the symref was seen on a `prepareForOSR` call at this point in the call chain, and the `definingSymRef` is the symref that actually holds that value at that point in the call chain.  So, for instance, in this case the defining map for #370 will be {#362}, and the defining map for #369 will be {#421}.

```
n363n       call  prepareForOSR[#55  helper Method]
n343n         loadaddr  vmThread[#409  MethodMeta]
n344n         iconst 3
n351n         aload  <temp slot 7>[#421  Auto]
n352n         iconst 369
n353n         iconst -1
n354n         aload  <auto slot 2>[#362  Auto]
n355n         iconst 370
n356n         iconst -1
```

If the symref doesn't have an entry in the defining map, it's some other auto or parm symref that still might be holding a reference to a candidate for stack allocation, so, if it's live, we add a reference to it for the call to `eaEscapeHelper` to ensure that any candidate it might contain can get heapified.  Before we added the code that checks the `induceDefiningMap`, `processSymbolReferences` would add references to all live auto and parm symrefs that contained addresses.

I'm not sure whether that handling of the "no entry in the defining map" case might be making things too pessimistic.

Reply:
Thanks @hzongaro that helps clarify the question a lot. What is the situation in a sym ref may not have an entry in the defining map ? i.e. is it some optimization that might cause that situation to occur ?

Reply:
Yes, it's likely a temporary that was introduced by an optimization later than OSRDefAnalysis - I'm not sure whether that's the only scenario.

Perhaps the handling of symrefs with no entry in the defining map is too pessimistic, and might result in unnecessary heapification of candidates or even reduce the likelihood of stack allocation in some cases?  I could do some follow up investigation/experimentation if you think that would be appropriate.

Reply:
I think it is probably fine because escape analysis will itself track what the relevant uses are for an object it is considering for stack allocation, i.e. if you have an extra load on a cold path (the induce call is almost certainly marked cold) it could lead to heapification but not a failure to stack allocate, I think.